app/services/trainees/set_academic_cycles.rb

```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+module Trainees
+  class SetAcademicCycles
+    include ServicePattern
+
+    DURATIONS = %w[hours months days years].freeze
+
+    def initialize(trainee:)
+      @trainee = trainee
+    end
+
+    def call
+      trainee.start_academic_cycle = start_academic_cycle
+      trainee.end_academic_cycle = end_academic_cycle
+
+      trainee.save!
+      trainee
+    end
+
+  private
+
+    attr_reader :trainee
+
+    delegate :commencement_date, :itt_start_date, :awarded_at,
+             :withdraw_date, :itt_end_date, :hesa_metadatum, :training_route,
+             to: :trainee
+
+    def start_academic_cycle
+      start_date.present? ? AcademicCycle.for_date(start_date) : AcademicCycle.current
+    end
+
+    def end_academic_cycle
+      AcademicCycle.for_date(end_date) if end_date.present?
```

Comment:
Should we be defaulting to something if end date is nil? As far as I can see this would only happen if we don't have a start date in which case we're defaulting to the current cycle for the start. Bit of a tricky one.

Reply:
@gpeng  Yeah.. I had a 'think' about this at the time:

- It only happens if the trainee is missing a start date.
- There are no non-draft trainees missing a start date, so this only applies to draft trainees.
- All draft trainees are missing a start date before the user completes the course details section so we'd be guessing for all trainees at some point.
- There is no end date filter on the draft page

so in my head all of the above mashed together made a conclusion of no default.